fspcc.1
.\" Manpage for fspcc.
.\" Contact v.maffione@gmail.com to correct errors or typos.
.TH "FSPCC" "1" "June 2013" "fspcc 1.1" "fspcc manual"

.SH NAME
fspcc \- An FSP compiler and LTS analysis tool

.SH SYNOPSIS
.B fspcc
[\fI-dpgasvh\fR] [\fI-S FILE\fR] [\fI-D NUM\fR] \fI-i FILE\fR [\fI-o FILE\fR]
.br
.B fspcc
[\fI-dpgasvh\fR] [\fI-S FILE\fR] [\fI-D NUM\fR] \fI-l FILE\fR


.SH DESCRIPTION
.PP
This command invokes an Finite State Process (FSP) compiler and a
Labelled Transition System (LTS) analysis tool.
.PP
The FSP language can be used to model concurrent systems, e.g.
multi-threaded/multi-process applications. An FSP model can be
compiled into a graph-based representation using an FSP compiler:
each FSP process is transformed into an equivalent LTS.
.PP
An LTS analysis tool is then able to:
.RS 3
.IP "   -"
detect concurrency problems in the concurrent system (e.g. deadlock,
starvation)
.RE
.RS 3
.IP "   -"
show that certain properties (or assertions) hold on the concurrent system
.RE
.RS 3
.IP "   -"
run a simulation (animation) of the concurrent system, in order to gain
confidence on the model
.RE
.PP
When invoked in the first form, you use the \fI-i\fR option to specify an
input file containing FSP definitions. The tool will compile all the FSP
models and produce an output file containing the corresponding compiled LTSs.
.PP
When invoked in the second form, you use the \fI-l\fR option to specify a
file containing a set of compiled LTSs, in order to use the LTS analysis
tool without compiling again.


.SH OPTIONS
.PP
\fB\-i\fR \fIPATHNAME\fR
.RS 3
Specifies an input file pathname containing FSP definitions to compile.
.RE

.PP
\fB\-o\fR \fIPATHNAME\fR
.RS 3
Specifies the output file pathname where compiled LTSs are stored. If this
option is not specified, the default output file name is 'output.lts'.
.RE

.PP
\fB\-l\fR \fIPATHNAME\fR
.RS 3
Specifies the input file pathname containing compiled LTSs.
.RE

.PP
\fB\-d\fR
.RS 3
Runs deadlock/error analysis on every compiled LTS. A list of deadlocks/errors
will be printed on the standard output. If there are no such problems, no
output is produced.
.RE

.PP
\fB\-p\fR
.RS 3
Runs all the specified progress checks on every compiled LTS. A list of
progress violations will be printed on the standard output. If there are no
such problems, no output is produced.
.RE

.PP
\fB\-g\fR
.RS 3
Outputs a GraphViz representation file corrisponding to every LTS.
For each process named 'P', a file called 'P.gv' is created.
.RE

.PP
\fB\-a\fR
.RS 3
The same as \fB\-dpg\fR.
.RE

.PP
\fB\-s\fR
.RS 3
Runs an LTS analysis interactive shell. The shell is run after compilation
(if any).
.RE

.PP
\fB\-S\fR \fIPATHNAME\fR
.RS 3
Runs the LTS analysis script specified by the pathname. The script is
run after the compilation (if any) but before invoking the interactive shell
(when \fB\-s\fR is specified).
.RE

.PP
\fB\-D\fR \fINUMBER\fR
.RS 3
Specifies the maximum accepted depth for process references. This is
mainly useful to avoid problems with poorly written recursive processes.
Example of a problematic process, which causes unterminated recursion:
.PP
.RS 4
P(K=3) = if (K == 0) then END else P(K-1);P(K-2);END.
.RE
.RE

.PP
\fB\-v\fR
.RS 3
Show versioning information.
.RE

.PP
\fB\-h\fR
.RS 3
Show the help.
.RE


.SH EXAMPLES
Here is a simple example of FSP input file (say in.fsp):
.PP
.RS 4
P = (a->b->P|c->d->END).
.br
Q = (t[i:1..2]-> (when (i>1) u->Q | v->Q)).
.br
||C = (P || Q).
.RE
.PP
Compile these FSP models and store the result into 'out.lts'
.PP
.RS 4
fspcc -i in.fsp -o out.lts
.RE
.PP
Now run all the analysis and produce GraphViz representations
.PP
.RS 4
fspcc -a -l out.lts
.RE
.PP
Alternatively, you can do the same thing with just one command
.PP
.RS 4
fspcc -a -i in.fsp -o out.lts
.RE
.PP
To launch an interactive shell
.PP
.RS 4
fspcc -s -i in.fsp
.RE
.PP
or, from a compiled LTS file
.PP
.RS 4
fspcc -s -l out.lts
.RE


.SH AUTHOR
Written by Vincenzo Maffione.

.SH REPORTING BUGS
Please report fspcc bugs to v.maffione@gmail.com.
.br
See fspcc project homepage: <http://www.bitbucket.org/vmaffione/fspc>

.SH COPYRIGHT
Copyright \(co 2013 Vincenzo Maffione.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

.SH SEE ALSO
\fBgraphviz\fR(7), \fBdot\fR(1)


fspcc.cpp
/*
 *  fspc - A Finite State Process compiler and LTS analyzer
 *
 *  Copyright (C) 2013  Vincenzo Maffione
 *  Email contact: <v.maffione@gmail.com>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


#include <iostream>
#include <cstdlib>
#include <cstring>
#include <unistd.h>

using namespace std;

/* Main FspDriver class, instantiating the parser. */
#include "driver.hpp"


void help()
{
    cout << "fspc - A Finite State Process compiler and LTS analisys tool.\n";
    cout << "USAGE: fspc [-dpgasSh] [-i FILE | -l FILE] [-o FILE]\n";
    cout << "   -i FILE : Specifies FILE as the input file containing "
            "FSP definitions.\n";
    cout << "   -l FILE : Specifies FILE as the input file containing "
            "compiled LTSs.\n";
    cout << "   -o FILE : Specifies the output FILE, e.g. the file that "
            "will contain the compiled LTSs.\n";
    cout << "   -d : Runs deadlock/error analysis on every FSP.\n";
    cout << "   -p : Runs all the specified progress verifications on "
            "every FSP.\n";
    cout << "   -g : Outputs a graphviz representation file of every FSP.\n";
    cout << "   -a : The same as '-d -p -g'\n";
    cout << "   -s : Runs an LTS analysis interactive shell\n";
    cout << "   -S FILE : Runs an LTS analysis script\n";
    cout << "   -D NUM : The maximum depth of process references accepted "
            "within a process definition (default is 1000)\n";
    cout << "   -v : Shows versioning information\n";
    cout << "   -h : Shows this help.\n";
}

#define GETOPT
#ifdef GETOPT
void process_args(CompilerOptions& co, int argc, char **argv)
{
    int ch;
    int il_options = 0;

    /* Set default values. */
    co.input_file = NULL;
    co.input_type = CompilerOptions::InputTypeFsp;
    co.output_file = NULL;
    co.deadlock = false;
    co.progress = false;
    co.graphviz = false;
    co.shell = false;
    co.script = false;
    co.max_reference_depth = 1000;

    while ((ch = getopt(argc, argv, "i:l:o:adpghsvS:D:")) != -1) {
	switch (ch) {
	    default:
		cout << "\n";
		help();
		exit(1);
		break;

	    case 'i':
		il_options++;
		co.input_file = optarg;
		co.input_type = CompilerOptions::InputTypeFsp;
		break;

	    case 'l':
		il_options++;
		co.input_file = optarg;
		co.input_type = CompilerOptions::InputTypeLts;
		break;

	    case 'o':
                co.output_file = optarg;
                break;

	    case 'a':
		co.deadlock = co.progress = co.graphviz = true;
		break;

	    case 'd':
		co.deadlock = true;
		break;

	    case 'p':
		co.progress = true;
		break;

	    case 'g':
		co.graphviz = true;
		break;

	    case 'h':
		help();
		exit(0);
		break;

	    case 's':
		co.shell = true;
		break;

	    case 'S':
		co.script = true;
		co.script_file = optarg;
		break;

            case 'D':
                co.max_reference_depth = atoi(optarg);
                break;

            case 'v':
                cout << "fspc 1.5 (January 2014)\n";
                cout << "Copyright 2013 Vincenzo Maffione\n";
                exit(0);
	}
    }

    if (il_options > 1) {
	cerr << "Error: Cannot specify more than one input file\n\n";
        help();
	exit(-1);
    }

    if (!co.input_file) {
        cerr << "Error: Missing input file\n\n";
        help();
        exit(-1);
    }

    if (!co.output_file && !co.shell && !co.script) {
        /* There's (normally) no point in running this program if one
           neither runs a shell nor a script nor generates a compiled
           output. Assume a default '-o' option. */
        co.output_file = "out.lts";
    }
}
#else	/* !GET_OPT */
void process_args(CompilerOptions& co, int argc, char **argv)
{
    int i = 1;
    int il_options = 0;

    /* Set default values. */
    co.input_file = "input.fsp";
    co.input_type = CompilerOptions::InputTypeFsp;
    co.output_file = "output.lts";
    co.deadlock = false;
    co.progress = false;
    co.graphviz = false;
    co.shell = false;

    /* Scan input arguments. */
    while (i < argc) {
	int len = strlen(argv[i]);
	const char *arg = argv[i];
	if (len != 2 || len == 2 && arg[0] != '-') {
	    cerr << "Error: Unrecognized option\n";
	    exit(-1);
	}
	switch (arg[1]) {
	    case 'a':
		co.deadlock = co.progress = co.graphviz = true;
		break;
	    case 'd':
		co.deadlock = true;
		break;
	    case 'p':
		co.progress = true;
		break;
	    case 'g':
		co.graphviz = true;
		break;
	    case 'h':
		help();
		exit(0);
		break;
	    case 'i':
	    case 'l':
	    case 'o':
		i++;
		if (i == argc) {
		    cerr << "Error: Expected filename after -"<< arg[1] 
			    << " option\n";
		    exit(-1);
		}
		switch (arg[1]) {
		    case 'i':
			il_options++;
			co.input_file = argv[i];
			break;
		    case 'l':
			il_options++;
			co.input_file = argv[i];
			co.input_type = CompilerOptions::InputTypeLts;
			break;
		    case 'o':
			co.output_file = argv[i];
			break;
		}
		break;
	    case 's':
		co.shell = true;
		break;

	    default:
		cerr << "Unrecognized option " << arg[1] << "\n";
		exit(-1);
	}
	i++;
    }

    if (il_options > 1) {
	cerr << "Error: Cannot specify both 'i' and 'l' options\n";
	exit(-1);
    }
}
#endif	/* !GET_OPT */


int main (int argc, char ** argv)
{
    FspDriver driver;
    CompilerOptions co;
    int ret;
    
    process_args(co, argc, argv);

    ret = driver.parse(co);

    return ret;
}
